Mathlib/Analysis/Asymptotics/ReflectsGrowth.lean

```diff
@@ -0,0 +1,267 @@
+
+## Keywords
+
+asymptotics
+-/
+
+open Function Filter Bornology
+open scoped Topology
+
+namespace Asymptotics
+
+section
+
+variable {ι α β E : Type*} [LinearOrder β] [NormedAddCommGroup E]
+
+/-- A function `g : ι → β` with linear ordered codomain
+*reflects growth* of a family of functions `f : ι → α → E` along a filter `l`, if
+
+- `f i =o[l] f j` iff `g i < g j`,
+- `f i =O[l] f j` iff `g i ≤ g j`,
+- none of `f i` is eventually equal to zero at `l`.
+
+The actual definition assumes only two implications
+-/
+structure ReflectsGrowth (f : ι → α → E) (g : ι → β) (l : Filter α) : Prop where
+  isLittleO_of_lt {i j} : g i < g j → f i =o[l] f j
+  isBigO_of_eq {i j} : g i = g j → f i =O[l] f j
+  frequently_ne {i} : ∃ᶠ x in l, f i x ≠ 0
+
+namespace ReflectsGrowth
+
+variable {f : ι → α → E} {g : ι → β} {l : Filter α} {i j : ι}
+
+lemma isBigO (h : ReflectsGrowth f g l) : f i =O[l] f j ↔ g i ≤ g j:=
+  ⟨fun hO ↦ not_lt.1 fun hlt ↦ (h.isLittleO_of_lt hlt).not_isBigO h.frequently_ne hO, fun hle ↦
+    hle.eq_or_lt.elim h.isBigO_of_eq fun hij ↦ (h.isLittleO_of_lt hij).isBigO⟩
+
+lemma isLittleO (h : ReflectsGrowth f g l) : f i =o[l] f j ↔ g i < g j := by
+  cases lt_or_le (g i) (g j) with
+  | inl hij => simp [hij, h.isLittleO_of_lt]
+  | inr hji => simp [hji.not_lt, (h.isBigO.2 hji).not_isLittleO h.frequently_ne]
+
+lemma isTheta (h : ReflectsGrowth f g l) : f i =Θ[l] f j ↔ g i = g j := by
+  simp only [IsTheta, h.isBigO, le_antisymm_iff]
+
+lemma congr_right {γ : Type*} [LinearOrder γ] {g' : ι → γ} (h : ReflectsGrowth f g l)
+    (hg : ∀ {i j}, g i < g j ↔ g' i < g' j) : ReflectsGrowth f g' l where
+  isLittleO_of_lt hij := h.isLittleO.2 <| hg.2 hij
+  isBigO_of_eq hij := h.isBigO.2 <| not_lt.1 <| mt hg.1 hij.not_gt
+  frequently_ne := h.frequently_ne
+
+lemma congr_left_isTheta {E' : Type*} [NormedAddCommGroup E'] {f' : ι → α → E'}
+    (h : ReflectsGrowth f g l) (hf : ∀ {i}, f i =Θ[l] f' i) : ReflectsGrowth f' g l where
+  isLittleO_of_lt := hf.isLittleO_congr_left.1 ∘ hf.isLittleO_congr_right.1 ∘ h.isLittleO.2
+  isBigO_of_eq := hf.isBigO_congr_left.1 ∘ hf.isBigO_congr_right.1 ∘ h.isBigO_of_eq
+  frequently_ne := (hf.eq_zero_iff.and_frequently h.frequently_ne).mono fun _ h ↦ mt h.1.2 h.2
+
+lemma congr_left {f' : ι → α → E} (h : ReflectsGrowth f g l) (hf : ∀ {i}, f i =ᶠ[l] f' i) :
+    ReflectsGrowth f' g l :=
+  h.congr_left_isTheta hf.isTheta
+
+protected lemma comp {ι' : Type*} (h : ReflectsGrowth f g l) (u : ι' → ι) :
+    ReflectsGrowth (f ∘ u) (g ∘ u) l :=
+  { h with }
+
+end ReflectsGrowth
+
+variable [Zero ι] [Zero β]
+
+/-- A zero homomorphism `g : ZeroHom ι β` with linear ordered codomain
+*reflects growth* of a family of functions `f : ι → α → E` along a filter `l`, if
+
+- `g` reflects growth of `f` along `l` in the sense of `Asymptotics.ReflectsGrowth`,
+- `g 0 = 0`,
+- `f 0 =Θ[l] 1`.
+-/
+structure ReflectsGrowth₀ (f : ι → α → E) (g : ZeroHom ι β) (l : Filter α)
```

Comment:
`_0` has the connotation of adding a zero to a multiplicative monoid, which is not what's happening here (I think?). What about
```suggestion
structure ReflectsGrowthZero (f : ι → α → E) (g : ZeroHom ι β) (l : Filter α)
```